File: code/java/data_structures_article-05.java
public class Data_structures_article_05 {
    /**
     * Core program of binary tree series algorithms
     */
    public Object solve(Object... args) {
        if (args == null || args.length != 1) {
            throw new IllegalArgumentException("Expected: (TreeNode root)");
        }

        Solution solution = new Solution();
        return solution.maxDepth((TreeNode) args[0]);
    }

    private static class Solution {

        int depth = 0;
        int res = 0;
        public int maxDepth(TreeNode root) {
            traverse(root);
            return res;
        }
        void traverse(TreeNode root) {
            if (root == null) {
                return;
            }
            depth++;
            res = Math.max(res, depth);
            traverse(root.left);
            traverse(root.right);
            depth--;
        }
    }
}

File: code/python/data_structures_article-05.py
"""Core program of binary tree series algorithms"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().maxDepth(*args, **kwargs)


class _Solution:
    def maxDepth(self, root: TreeNode) -> int:
        self.res = 0
        self.traverse(root, 0)
        return self.res
    def traverse(self, root: TreeNode, depth: int) -> None:
        if not root:
            return
        depth += 1
        self.res = max(self.res, depth)
        self.traverse(root.left, depth)
        self.traverse(root.right, depth)
        depth -= 1
class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if not root:
            return 0
        leftMax = self.maxDepth(root.left)
        rightMax = self.maxDepth(root.right)
        return 1 + max(leftMax, rightMax)
